molecule/provisioners/libvirtprovisioner.py

```diff
@@ -0,0 +1,923 @@
+
+        Since molecule-created instances are ephemeral, always undefine the libvirt domain and its storage volume.
+        """
+        domains = self._libvirt.listAllDomains()
+        pool = self._up_pool()
+        for instance in self.instances:
+            utilities.print_info("\tDestroying libvirt instance {} ...".format(
+                instance['name']))
+            dom_found = False
+            for dom in domains:
+                if not dom_found:
+                    if dom.name() == instance['name']:
+                        dom_found = True
+                        try:
+                            dom.destroy()
+                        except:
+                            pass
+                        try:
+                            self._destroy_volume(pool, instance)
+                        except:
+                            pass
+                        try:
+                            dom.undefine()
+                        except:
+                            pass
+                            # TODO: Consider whether these should really cause molecule to exit fail?
+                        utilities.print_success(
+                            '\tDestroyed and undefined libvirt instance {}'.format(
+                                instance['name']))
+            # TODO: Consider whether to destroy/undefine molecule networks if they are no longer used
+        return True
+
+    def status(self):
+        states = ['no state', 'running', 'blocked', 'paused', 'being shutdown',
+                  'shutoff', 'crashed', 'pmsuspended']
+        Status = collections.namedtuple('Status',
+                                        ['name', 'state', 'provider'])
+        status_list = []
+        domains = self._libvirt.listAllDomains()
+        for instance in self.instances:
+            ins_found = False
+            for dom in domains:
+                if not ins_found:
+                    if dom.name() == instance['name']:
+                        ins_found = True
+                        status_list.append(
+                            Status(
+                                name=instance['name'],
+                                state=states[dom.info()[0]],
+                                provider='Libvirt'))
+            if not ins_found:
+                status_list.append(
+                    Status(
+                        name=instance['name'],
+                        state='undefined',
+                        provider='Libvirt'))
+
+        return status_list
+
+    def _lease_ip(self, interface):
+        """
+        Lease IP address
+
+        :param: an interface dict that has at least: {'mac': MAC address, 'network_name': (libvirt) name of network}
+        :return: an IP address
+        """
+        delay = 1
+        max_count = 6
+        net = self._libvirt.networkLookupByName(interface['network_name'])
+        count = 0
+        while not 'ip' not in interface:
```

Comment:
I believe this loop is more complicated than what it should be:
- the while condition is weird because the looping is limited by the `count` and you `return` when you find your value
- there is too much sleeping

Here is my suggestion (100% untested code but it should give you an idea of what I'm suggesting):
```
for iteration in range(10):
    leases = net.DHCPLeases()
    matching_ipaddr = next((lease['ipaddr'] for lease in leases if lease['mac'] == interface['mac']), None)
    if matching_ipaddr is not None:
        return matching_ipaddr
    time.sleep(delay + iteration)
```